codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```
@@ -464,4 +444,125 @@ public Http2Settings pollSentSettings() {
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            data.release();
+            promise.tryFailure(cause);
+        }
+
+        @Override
+        public boolean write(int allowedBytes) {
+            if (allowedBytes == 0 && size() != 0) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return false;
+            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            try {
+                int bytesWritten = 0;
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    ByteBuf toWrite;
+                    // Let data consume the frame before padding.
+                    if (data.readableBytes() == 0) {
+                        toWrite = Unpooled.EMPTY_BUFFER;
+                    } else  if (data.readableBytes() > allowedFrameSize) {
+                        toWrite = data.readSlice(Math.min(data.readableBytes(), allowedFrameSize)).retain();
+                    } else {
+                        toWrite = data;
+                    }
+                    int writeablePadding = Math.min(allowedFrameSize - toWrite.readableBytes(), padding);
+                    padding -= writeablePadding;
+                    bytesWritten += toWrite.readableBytes() + writeablePadding;
+                    frameWriter().writeData(ctx, streamId, toWrite, writeablePadding,
+                            size == bytesWritten && endOfStream,
+                            size == bytesWritten ? promise : ctx.voidPromise());
+                } while (size != bytesWritten && allowedBytes > bytesWritten);
+                size -= bytesWritten;
+                return true;
+            } catch (Throwable e) {
+                promise.tryFailure(e);
+                ReferenceCountUtil.safeRelease(data);
```

I was suggesting using `release()`.  I'm not to keen on silently discarding (just log statement) exceptions.